Add ability to reconnect based on number of failed ping requests

davidertel opened the pull request:
When running jitsi videobridge in amazon, we connect to our XMPP servers through an ELB. When deploys of the XMPP server happen, the nodes are terminated but the connection to the ELB still stays up, even though the nodes behind it have been removed. This causes the videobridge to think that it still has a connection to the XMPP server. Although the connection still appears valid, nothing is there to respond to pings or pubsub messages so the videobridge gets into a bad state. Forcing the underlying connection to reset based on failed ping requests helped to solve this for us.

The jitsi videobridge pom.xml will need to be updated to use the new version of Jicoco once this is merged into master. I'm not sure of the proper way to make that happen together so please advise here. Also did not see any documentation for jicoco and would be happy to start some but it felt like since this is used in the videobridge that the documentation should go there. Please advise as to where I should document this new configuration property.

The new configuration property can be set in the sip-comminicator.properties by adding the property _configPropertiesBase_.RECONNECT_ON_PING_FAILURES and setting it to true. To maintain the original behavior is treated as false by default. Setting this value will cause the connection to be reset after the number of pings set in the ping threshold fail.

Commit:
Add ability to reconnect based on number of failed ping requests. This can be configured in the sip-comminicator.properties by setting the property configPropertiesBase.RECONNECT_ON_PING_FAILURES to true. To maintain the original behavior this is set to false by default. Setting this value will cause the connection to be reset after the number of pings set in the ping threshold fail.

davidertel commented:
bump... does this repo get love anymore?

Commit:
The connection wasn't being completely reset, once the component is removed from the ComponentManager the connection is disconnected and properly reset

Commit:
revert imports back to *

Commit:
add documentation for new ping property

gpolitis commented on src/main/java/org/jitsi/xmpp/component/ComponentBase.java:
```diff
@@ -271,6 +287,25 @@ public boolean isConnectionAlive()
     }
 
+    /**
+     * Restart the component connection if we have exceeded the threshold for ping failures
```
Could you please break this comment to fit in 80 chars.

gpolitis commented on src/main/java/org/jitsi/xmpp/component/ComponentBase.java:
```diff
@@ -56,6 +56,11 @@
     private final static String PING_THRESHOLD_PNAME = "PING_THRESHOLD";
 
     /**
+     * The name of the property used to determine if we should reconnect on ping failures
```
Could you please break this comment to fit in 80 chars.

gpolitis commented on src/main/java/org/jitsi/xmpp/component/ComponentBase.java:
```diff
@@ -56,6 +56,11 @@
     /**
+     * The name of the property used to determine if we should reconnect on ping failures
+     */
+    private final static String RECONNECT_ON_PING_FAILURES = "RECONNECT_ON_PING_FAILURES";
```
Could you please break this comment to fit in 80 chars.

gpolitis commented on src/main/java/org/jitsi/xmpp/component/ComponentBase.java:
```diff
@@ -113,6 +118,12 @@
     private int pingFailures = 0;
 
     /**
+     * If we exceed the maximum number of ping failures should we trigger a reconnect
```
Could you please break this comment to fit in 80 chars.

gpolitis commented on src/main/java/org/jitsi/xmpp/component/ComponentBase.java:
```diff
@@ -271,6 +287,25 @@ public boolean isConnectionAlive()
+            logger.error("Failed to restart xmpp component");
+        }
+
+//        this.start();
```
Can we remove this please?

gpolitis commented:
Thank you very much for the detailed description. Reading it has helped me understand why we need this. Could you please add this text (or something similar) in the commit description or inside the code?

Other than that, the code looks good to me.

Commit:
When running jitsi videobridge in amazon, we connect to our XMPP servers through an ELB. When deploys of the XMPP server happen, the nodes are terminated but the connection to the ELB still stays up, even though the nodes behind it have been removed. This causes the videobridge to think that it still has a connection to the XMPP server. Although the connection still appears valid, nothing is there to respond to pings or pubsub messages so the videobridge gets into a bad state. Forcing the underlying connection to reset based on failed ping requests helped to solve this for us.

The jitsi videobridge pom.xml will need to be updated to use the new version of Jicoco once this is merged into master. I'm not sure of the proper way to make that happen together so please advise here. Also did not see any documentation for jicoco and would be happy to start some but it felt like since this is used in the videobridge that the documentation should go there. Please advise as to where I should document this new configuration property.

The new configuration property can be set in the sip-comminicator.properties by adding the property configPropertiesBase.RECONNECT_ON_PING_FAILURES and setting it to true. To maintain the original behavior is treated as false by default. Setting this value will cause the connection to be reset after the number of pings set in the ping threshold fail.

davidertel commented:
Updated the commit with your comments, thank you for reviewing.

Commit:
When running jitsi videobridge in amazon, we connect to our XMPP servers through an ELB. When deploys of the XMPP server happen, the nodes are terminated but the connection to the ELB still stays up, even though the nodes behind it have been removed. This causes the videobridge to think that it still has a connection to the XMPP server. Although the connection still appears valid, nothing is there to respond to pings or pubsub messages so the videobridge gets into a bad state. Forcing the underlying connection to reset based on failed ping requests helped to solve this for us.

The jitsi videobridge pom.xml will need to be updated to use the new version of Jicoco once this is merged into master. I'm not sure of the proper way to make that happen together so please advise here. Also did not see any documentation for jicoco and would be happy to start some but it felt like since this is used in the videobridge that the documentation should go there. Please advise as to where I should document this new configuration property.

The new configuration property can be set in the sip-comminicator.properties by adding the property configPropertiesBase.RECONNECT_ON_PING_FAILURES and setting it to true. To maintain the original behavior is treated as false by default. Setting this value will cause the connection to be reset after the number of pings set in the ping threshold fail.

davidertel commented:
@gpolitis anything else you would like me to update on this PR? Thank you again for reviewing it

davidertel commented:
bump?

gpolitis commented:
Hi @davidertel, I apologize for the late response. I'm sorry to bother you with this, but could you please make sure that your code does not go beyond 80 chars? You can configure IntelliJ IDEA (or any other tool) to show you a ruler. Thanks very much and again, I'm sorry for bothering with this and for taking so long to reply.

Commit:
reset pom version to 1.0

jitsi-jenkins commented:
Hi, thanks for your contribution!
If you haven't already done so, could you please make sure you sign our CLA (https://jitsi.org/icla for individuals and https://jitsi.org/ccla for corporations)? We would unfortunately be unable to merge your patch unless we have that piece :(.

davidertel commented:
Signed that several months ago. You should have it on record

> On Nov 1, 2016, at 4:10 PM, jitsi-jenkins <notifications@github.com> wrote:
> 
> Hi, thanks for your contribution!
> If you haven't already done so, could you please make sure you sign our CLA (https://jitsi.org/icla for individuals and https://jitsi.org/ccla for corporations)? We would unfortunately be unable to merge your patch unless we have that piece :(.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

bgrozev commented:
We do, sorry for the noise. 

Jenkins: add to whitelist.

Commit:
Merge branch 'master' of https://github.com/jitsi/jicoco

Commit:
don't let code go past 80 characters because the world still loves tiny screens??

davidertel commented:
merged in the latest from master to update this PR, reduced the width of the code to 80 characters... who has a screen that only supports 80 characters? email me and I'll send you a nice widescreen :-)

damencho commented:
Yep, there were a lot of discussions on the 80 chars online and offline, but one big + for that one is mobile. Also you should count it 160, normally you look diffs of two files side by side.

davidertel commented:
You guys must be a lot more adventurous than me, skydiving with your mobile phone while reviewing PRs, that's next level stuff. Anything else that you would like me to update before merging?

davidertel commented:
Anything further needed for this PR before merging?

Commit:
Merge branch 'master' of https://github.com/jitsi/jicoco

dy93 commented:
Hi, I encounter the same problem when using NLB to connect to xmpp server for jvb.
Any one can help to merge this PR?
